kvs/client/python/anna/ipc_client.py

```diff
@@ -12,59 +12,167 @@
+
+        self.put_push_socket = self.context.socket(zmq.PUSH)
+        self.put_push_socket.connect(PUT_REQUEST_ADDR)
+
+        self.get_pull_socket = self.context.socket(zmq.PULL)
+        self.get_pull_socket.bind(GET_RESPONSE_ADDR_TEMPLATE % thread_id)
+        self.get_pull_socket.RCVTIMEO = TIMEOUT
 
-        self.put_socket = self.context.socket(zmq.REQ)
-        self.put_socket.connect(PUT_ADDR)
+        self.put_pull_socket = self.context.socket(zmq.PULL)
+        self.put_pull_socket.bind(PUT_RESPONSE_ADDR_TEMPLATE % thread_id)
+        self.put_pull_socket.RCVTIMEO = TIMEOUT
 
-    def get(self, key, ltype):
+    def get(self, refs):
         request = KeyRequest()
         request.type = GET
 
-        tp = request.tuples.add()
-        tp.lattice_type = ltype
-        tp.key = key
+        deserialize = {}
+
+        for ref in refs:
+            tp = request.tuples.add()
+            tp.key = ref.key
+            tp.lattice_type = ref.obj_type
+
+            deserialize[ref.key] = ref.deserialize
 
-        self.get_socket.send(request.SerializeToString())
+            if ref.obj_type == CROSSCAUSAL:
+                print("Error: found cross causal lattice type in non causal mode!")
 
-        resp = KeyResponse()
-        resp.ParseFromString(self.get_socket.recv())
+        request.response_address = GET_RESPONSE_ADDR_TEMPLATE % thread_id
 
-        tp = resp.tuples[0]
+        self.get_push_socket.send(request.SerializeToString())
 
-        if tp.error == 1:
-            print('Key %s does not exist!' % (key))
+        try:
+            msg = self.get_pull_socket.recv()
+        except zmq.ZMQError as e:
+            if e.errno == zmq.EAGAIN:
+                print("request timed out")
+                return None
+            else:
+                print("unknown zmq error")
+                return None
+        else:
+            kv_pairs = {}
+            resp = KeyResponse()
+            resp.ParseFromString(msg)
+
+            for tp in resp.tuples:
+                if tp.error == 1:
+                    print('Key %s does not exist!' % (key))
+                    return None
+
+                if tp.lattice_type == LWW:
+                    val = LWWValue()
+                    val.ParseFromString(tp.payload)
+
+                    if deserialize[tp.key]:
+                        kv_pairs[tp.key] = deserialize_val(val.value)
+                    else:
+                        kv_pairs[tp.key] = val.value
+                elif tp.lattice_type == SET:
+                    res = set()
+
+                    val = SetValue()
+                    val.ParseFromString(tp.payload)
+
+                    for v in val.values:
+                        res.add(v)
+
+                    # for set lattice, just pick the first one to return for now
+                    # we are not using it anyway..
+                    if deserialize[tp.key]:
+                        kv_pairs[tp.key] = deserialize_val(res[0])
```

Comment:
This is not going to work. The IPC client is in a completely different package and doesn't have access to the `deserialize_val` method -- which is not a KVS construct anyway. This should be in the function execution code.

Reply:
got it. moved this to executor...